File: Topological Sorting/14907번 - 프로젝트 스케줄링/14907.cpp
#include <iostream>
#include <vector>
#include <sstream>
#include <queue>
#include <algorithm>

using namespace std;

#define LIM 27

int degree[LIM] = { 0, }, t[LIM] = { 0, };
vector<int> order[LIM];

vector<string> split(string st, char del) {
    istringstream iss(st);
    string buffer;
    
    vector<string> res;
    
    while(getline(iss, buffer, del)) {
        res.push_back(buffer);
    }
    
    return res;
}

void input() {
    string str;
    
    while(getline(cin, str)) {
        vector<string> res = split(str, ' ');
        int alpha = res[0][0] - 'A';
        
        t[alpha] = stoi(res[1]);
        
        if(res.size() > 2) {
            for(auto &c : res[2]) {
                order[c - 'A'].push_back(alpha);
                degree[alpha]++;
            }
        }
    }
}

void solve() {
    int min_time[LIM] = { 0, };
    
    queue<int> Q;
    for(int i = 0; i < LIM; i++) {
        if(degree[i] == 0 && t[i]) {
            Q.push(i);
            min_time[i] = t[i];
        }
    }
    
    while(!Q.empty()) {
        int now = Q.front(); Q.pop();

        for(int &nxt : order[now]) {
            min_time[nxt] = max(min_time[nxt], min_time[now] + t[nxt]);
            
            if(--degree[nxt] == 0) Q.push(nxt);
        }
    }
    
    cout << *max_element(min_time, min_time + LIM) << '\n';
}

int main() {
    ios::sync_with_stdio(0); 
    cin.tie(0);
    
    input();
    solve();
    
    return 0;
}

File: Topological Sorting/14907번 - 프로젝트 스케줄링/14907.py
import sys
from collections import deque

input = sys.stdin.readline

LIM = 27

degree = [0] * LIM
time = [0] * LIM
project = [[] for _ in range(LIM)]

while True:
    try:
        tmp = input().split()
        alpha = ord(tmp[0]) - ord('A')
        time[alpha] = int(tmp[1])
        
        if len(tmp) > 2:
            for c in tmp[2]:
                project[ord(c) - ord('A')].append(alpha)
            degree[alpha] = len(tmp[2])
    except:
        break

def solve():
    q = deque()
    min_time = [0] * LIM
    
    for i in range(LIM):
        if degree[i] == 0 and time[i]:
            q.append(i)
            min_time[i] = time[i]
    
    while q:
        now = q.popleft()
        
        for nxt in project[now]:
            min_time[nxt] = max(min_time[nxt], min_time[now] + time[nxt])
            
            degree[nxt] -= 1
            if degree[nxt] == 0:
                q.append(nxt)
                
    print(max(min_time))
        

if __name__ == '__main__':
    solve()